<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.433">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="description" content="bite sized functions for processing adata">

<title>degex - adata</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<link rel="stylesheet" href="styles.css">
<meta property="og:title" content="degex - adata">
<meta property="og:description" content="bite sized functions for processing adata">
<meta property="og:site-name" content="degex">
<meta name="twitter:title" content="degex - adata">
<meta name="twitter:description" content="bite sized functions for processing adata">
<meta name="twitter:card" content="summary">
</head>

<body class="nav-sidebar floating nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">degex</span>
    </a>
  </div>
        <div class="quarto-navbar-tools ms-auto">
</div>
          <div id="quarto-search" class="" title="Search"></div>
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./adata.html">adata</a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation floating overflow-auto">
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">DEtect Gene EXpression (degex)</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./static.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">static</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./types.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">types</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./utils.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">utils</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./adata.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">adata</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./plots.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Plots</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./preprocessing.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">preprocessing</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#adjust-anndata-var" id="toc-adjust-anndata-var" class="nav-link active" data-scroll-target="#adjust-anndata-var">Adjust AnnData <code>var</code></a></li>
  <li><a href="#score_doublets" id="toc-score_doublets" class="nav-link" data-scroll-target="#score_doublets">score_doublets</a></li>
  <li><a href="#remove_mitochondrial_genes" id="toc-remove_mitochondrial_genes" class="nav-link" data-scroll-target="#remove_mitochondrial_genes">remove_mitochondrial_genes</a></li>
  <li><a href="#set_var_names_as_gene_ids" id="toc-set_var_names_as_gene_ids" class="nav-link" data-scroll-target="#set_var_names_as_gene_ids">set_var_names_as_gene_ids</a></li>
  <li><a href="#set_gene_symbol_as_var_names" id="toc-set_gene_symbol_as_var_names" class="nav-link" data-scroll-target="#set_gene_symbol_as_var_names">set_gene_symbol_as_var_names</a></li>
  <li><a href="#annotations" id="toc-annotations" class="nav-link" data-scroll-target="#annotations">Annotations</a></li>
  <li><a href="#add_gene_annotations" id="toc-add_gene_annotations" class="nav-link" data-scroll-target="#add_gene_annotations">add_gene_annotations</a></li>
  <li><a href="#stack-and-batch" id="toc-stack-and-batch" class="nav-link" data-scroll-target="#stack-and-batch">Stack and Batch</a></li>
  <li><a href="#stack" id="toc-stack" class="nav-link" data-scroll-target="#stack">stack</a></li>
  <li><a href="#stack_batchs" id="toc-stack_batchs" class="nav-link" data-scroll-target="#stack_batchs">stack_batchs</a></li>
  <li><a href="#quality-control" id="toc-quality-control" class="nav-link" data-scroll-target="#quality-control">Quality Control</a>
  <ul class="collapse">
  <li><a href="#calc_qc_stats" id="toc-calc_qc_stats" class="nav-link" data-scroll-target="#calc_qc_stats">calc_qc_stats</a></li>
  <li><a href="#make_var_starts_with" id="toc-make_var_starts_with" class="nav-link" data-scroll-target="#make_var_starts_with">make_var_starts_with</a></li>
  <li><a href="#var_starts_with_pattern" id="toc-var_starts_with_pattern" class="nav-link" data-scroll-target="#var_starts_with_pattern">var_starts_with_pattern</a></li>
  <li><a href="#apply_filter_by_cutoffs" id="toc-apply_filter_by_cutoffs" class="nav-link" data-scroll-target="#apply_filter_by_cutoffs">apply_filter_by_cutoffs</a></li>
  <li><a href="#filter_by_cutoffs" id="toc-filter_by_cutoffs" class="nav-link" data-scroll-target="#filter_by_cutoffs">filter_by_cutoffs</a></li>
  </ul></li>
  <li><a href="#normalization" id="toc-normalization" class="nav-link" data-scroll-target="#normalization">Normalization</a>
  <ul class="collapse">
  <li><a href="#sqrt_library_size_normalize" id="toc-sqrt_library_size_normalize" class="nav-link" data-scroll-target="#sqrt_library_size_normalize">sqrt_library_size_normalize</a></li>
  <li><a href="#add_gene_detection_layer" id="toc-add_gene_detection_layer" class="nav-link" data-scroll-target="#add_gene_detection_layer">add_gene_detection_layer</a></li>
  <li><a href="#add_prenormalization_layer" id="toc-add_prenormalization_layer" class="nav-link" data-scroll-target="#add_prenormalization_layer">add_prenormalization_layer</a></li>
  <li><a href="#load_human_genes" id="toc-load_human_genes" class="nav-link" data-scroll-target="#load_human_genes">load_human_genes</a></li>
  <li><a href="#score_genes_cell_cycle_with_batch_mean_center_data" id="toc-score_genes_cell_cycle_with_batch_mean_center_data" class="nav-link" data-scroll-target="#score_genes_cell_cycle_with_batch_mean_center_data">score_genes_cell_cycle_with_batch_mean_center_data</a></li>
  <li><a href="#add_batch_mean_center_layer" id="toc-add_batch_mean_center_layer" class="nav-link" data-scroll-target="#add_batch_mean_center_layer">add_batch_mean_center_layer</a></li>
  </ul></li>
  <li><a href="#highly-variable-genes" id="toc-highly-variable-genes" class="nav-link" data-scroll-target="#highly-variable-genes">Highly Variable Genes</a>
  <ul class="collapse">
  <li><a href="#select_hvg_per_batch" id="toc-select_hvg_per_batch" class="nav-link" data-scroll-target="#select_hvg_per_batch">select_hvg_per_batch</a></li>
  </ul></li>
  <li><a href="#annotations-1" id="toc-annotations-1" class="nav-link" data-scroll-target="#annotations-1">Annotations</a>
  <ul class="collapse">
  <li><a href="#add_mouse_tfs_from_csv" id="toc-add_mouse_tfs_from_csv" class="nav-link" data-scroll-target="#add_mouse_tfs_from_csv">add_mouse_tfs_from_csv</a></li>
  <li><a href="#add_human_tfs_from_csv" id="toc-add_human_tfs_from_csv" class="nav-link" data-scroll-target="#add_human_tfs_from_csv">add_human_tfs_from_csv</a></li>
  <li><a href="#add_tf_annotations_from_csv" id="toc-add_tf_annotations_from_csv" class="nav-link" data-scroll-target="#add_tf_annotations_from_csv">add_tf_annotations_from_csv</a></li>
  </ul></li>
  <li><a href="#normalization-1" id="toc-normalization-1" class="nav-link" data-scroll-target="#normalization-1">Normalization</a>
  <ul class="collapse">
  <li><a href="#subset_markers" id="toc-subset_markers" class="nav-link" data-scroll-target="#subset_markers">subset_markers</a></li>
  <li><a href="#zscore_markers_in_layer" id="toc-zscore_markers_in_layer" class="nav-link" data-scroll-target="#zscore_markers_in_layer">zscore_markers_in_layer</a></li>
  </ul></li>
  <li><a href="#embeddings" id="toc-embeddings" class="nav-link" data-scroll-target="#embeddings">Embeddings</a>
  <ul class="collapse">
  <li><a href="#pca" id="toc-pca" class="nav-link" data-scroll-target="#pca">PCA</a></li>
  <li><a href="#run_pca_on_hvg" id="toc-run_pca_on_hvg" class="nav-link" data-scroll-target="#run_pca_on_hvg">run_pca_on_hvg</a></li>
  <li><a href="#run_pca" id="toc-run_pca" class="nav-link" data-scroll-target="#run_pca">run_pca</a></li>
  <li><a href="#phate" id="toc-phate" class="nav-link" data-scroll-target="#phate">PHATE</a></li>
  <li><a href="#run_phate_on_hvg" id="toc-run_phate_on_hvg" class="nav-link" data-scroll-target="#run_phate_on_hvg">run_phate_on_hvg</a></li>
  <li><a href="#run_phate_using_g" id="toc-run_phate_using_g" class="nav-link" data-scroll-target="#run_phate_using_g">run_phate_using_g</a></li>
  <li><a href="#magic" id="toc-magic" class="nav-link" data-scroll-target="#magic">MAGIC</a></li>
  <li><a href="#run_magic" id="toc-run_magic" class="nav-link" data-scroll-target="#run_magic">run_magic</a></li>
  </ul></li>
  </ul>
<div class="toc-actions"><div><i class="bi bi-github"></i></div><div class="action-links"><p><a href="https://github.com/dsm-72/degex/issues/new" class="toc-action">Report an issue</a></p></div></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">adata</h1>
</div>

<div>
  <div class="description">
    bite sized functions for processing adata
  </div>
</div>


<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<!-- WARNING: THIS FILE WAS AUTOGENERATED! DO NOT EDIT! -->
<section id="adjust-anndata-var" class="level3">
<h3 class="anchored" data-anchor-id="adjust-anndata-var">Adjust AnnData <code>var</code></h3>
<hr>
<p><a href="https://github.com/dsm-72/degex/blob/main/degex/adata.py#L111" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="score_doublets" class="level3">
<h3 class="anchored" data-anchor-id="score_doublets">score_doublets</h3>
<blockquote class="blockquote">
<pre><code> score_doublets (adata:anndata._core.anndata.AnnData, plot:bool=False,
                 **kwargs)</code></pre>
</blockquote>
<hr>
<p><a href="https://github.com/dsm-72/degex/blob/main/degex/adata.py#L94" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="remove_mitochondrial_genes" class="level3">
<h3 class="anchored" data-anchor-id="remove_mitochondrial_genes">remove_mitochondrial_genes</h3>
<blockquote class="blockquote">
<pre><code> remove_mitochondrial_genes (adata:anndata._core.anndata.AnnData)</code></pre>
</blockquote>
<p>Filters <code>adata</code> by colums not in <code>adata.var.mito</code>.</p>
<table class="table">
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>adata</td>
<td>AnnData</td>
<td>AnnData to process</td>
</tr>
<tr class="even">
<td><strong>Returns</strong></td>
<td><strong>AnnData</strong></td>
<td><strong>for function chaining</strong></td>
</tr>
</tbody>
</table>
<hr>
<p><a href="https://github.com/dsm-72/degex/blob/main/degex/adata.py#L77" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="set_var_names_as_gene_ids" class="level3">
<h3 class="anchored" data-anchor-id="set_var_names_as_gene_ids">set_var_names_as_gene_ids</h3>
<blockquote class="blockquote">
<pre><code> set_var_names_as_gene_ids (adata:anndata._core.anndata.AnnData)</code></pre>
</blockquote>
<p>Adds <code>gene_ids</code> to <code>adata.var</code>.</p>
<table class="table">
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>adata</td>
<td>AnnData</td>
<td>AnnData to process</td>
</tr>
<tr class="even">
<td><strong>Returns</strong></td>
<td><strong>AnnData</strong></td>
<td><strong>for function chaining</strong></td>
</tr>
</tbody>
</table>
<hr>
<p><a href="https://github.com/dsm-72/degex/blob/main/degex/adata.py#L57" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="set_gene_symbol_as_var_names" class="level3">
<h3 class="anchored" data-anchor-id="set_gene_symbol_as_var_names">set_gene_symbol_as_var_names</h3>
<blockquote class="blockquote">
<pre><code> set_gene_symbol_as_var_names (adata:anndata._core.anndata.AnnData)</code></pre>
</blockquote>
<p>Enfoces that <code>adata.var</code> names are unique and adds <code>gene_symbol</code> to <code>adata.var</code>.</p>
<table class="table">
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>adata</td>
<td>AnnData</td>
<td>AnnData to process</td>
</tr>
<tr class="even">
<td><strong>Returns</strong></td>
<td><strong>AnnData</strong></td>
<td><strong>for function chaining</strong></td>
</tr>
</tbody>
</table>
</section>
<section id="annotations" class="level3">
<h3 class="anchored" data-anchor-id="annotations">Annotations</h3>
<pre><code>/opt/hostedtoolcache/Python/3.11.4/x64/lib/python3.11/site-packages/fastcore/docscrape.py:225: UserWarning: Unknown section Notes
  else: warn(msg)</code></pre>
<hr>
<p><a href="https://github.com/dsm-72/degex/blob/main/degex/adata.py#L141" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="add_gene_annotations" class="level3">
<h3 class="anchored" data-anchor-id="add_gene_annotations">add_gene_annotations</h3>
<blockquote class="blockquote">
<pre><code> add_gene_annotations (adata:anndata._core.anndata.AnnData,
                       annotation_file:str)</code></pre>
</blockquote>
<p>Reads CSV and adds Ensembl information to <code>adata.var.index</code></p>
<table class="table">
<colgroup>
<col style="width: 9%">
<col style="width: 38%">
<col style="width: 52%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>adata</td>
<td>AnnData</td>
<td>AnnData to process</td>
</tr>
<tr class="even">
<td>annotation_file</td>
<td>str</td>
<td>full path to csv file. Assumed to have a column<br>called <code>Ensembl</code>.</td>
</tr>
<tr class="odd">
<td><strong>Returns</strong></td>
<td><strong>AnnData</strong></td>
<td><strong>for function chaining</strong></td>
</tr>
</tbody>
</table>
</section>
<section id="stack-and-batch" class="level3">
<h3 class="anchored" data-anchor-id="stack-and-batch">Stack and Batch</h3>
<hr>
<p><a href="https://github.com/dsm-72/degex/blob/main/degex/adata.py#L186" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="stack" class="level3">
<h3 class="anchored" data-anchor-id="stack">stack</h3>
<blockquote class="blockquote">
<pre><code> stack (*adatas:List[anndata._core.anndata.AnnData], key:str='_idx',
        replace:Optional[Dict[str,str]]=None,
        replace_key:Optional[str]=None, print_counts:bool=False)</code></pre>
</blockquote>
<pre><code>/opt/hostedtoolcache/Python/3.11.4/x64/lib/python3.11/site-packages/fastcore/docscrape.py:225: UserWarning: Unknown section Examples
  else: warn(msg)</code></pre>
<hr>
<p><a href="https://github.com/dsm-72/degex/blob/main/degex/adata.py#L205" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="stack_batchs" class="level3">
<h3 class="anchored" data-anchor-id="stack_batchs">stack_batchs</h3>
<blockquote class="blockquote">
<pre><code> stack_batchs (*adatas:List[anndata._core.anndata.AnnData],
               idx_to_time:dict, idx_to_batch:dict=None,
               batch_to_timepoint:dict=None, print_counts:bool=False)</code></pre>
</blockquote>
<p>Concatenates the each one of the <code>adata</code>s in<br>
<code>adatas</code> to a single <code>adata</code> instance.</p>
<table class="table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>adatas</td>
<td>List</td>
<td></td>
<td></td>
</tr>
<tr class="even">
<td>idx_to_time</td>
<td>dict</td>
<td></td>
<td>map of {int|str: str} for indicies to human friendly <br>time values.</td>
</tr>
<tr class="odd">
<td>idx_to_batch</td>
<td>dict</td>
<td>None</td>
<td>map of {int|str: str} for indicies to batch</td>
</tr>
<tr class="even">
<td>batch_to_timepoint</td>
<td>dict</td>
<td>None</td>
<td>map of {str: int} to map each batch (from <code>idx_to_time</code> or <code>idx_to_batch</code>) to a timepoint</td>
</tr>
<tr class="odd">
<td>print_counts</td>
<td>bool</td>
<td>False</td>
<td>Whether or not to print batched value counts.</td>
</tr>
<tr class="even">
<td><strong>Returns</strong></td>
<td><strong>AnnData</strong></td>
<td></td>
<td><strong>Concatenated adata.</strong></td>
</tr>
</tbody>
</table>
</section>
<section id="quality-control" class="level2">
<h2 class="anchored" data-anchor-id="quality-control">Quality Control</h2>
<hr>
<p><a href="https://github.com/dsm-72/degex/blob/main/degex/adata.py#L280" target="_blank" style="float:right; font-size:smaller">source</a></p>
<section id="calc_qc_stats" class="level3">
<h3 class="anchored" data-anchor-id="calc_qc_stats">calc_qc_stats</h3>
<blockquote class="blockquote">
<pre><code> calc_qc_stats (adata:anndata._core.anndata.AnnData,
                qc_vars:Dict[str,tuple]={'mito': ('mt-',), 'ribo': ('rps',
                'rpl')})</code></pre>
</blockquote>
<hr>
<p><a href="https://github.com/dsm-72/degex/blob/main/degex/adata.py#L276" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="make_var_starts_with" class="level3">
<h3 class="anchored" data-anchor-id="make_var_starts_with">make_var_starts_with</h3>
<blockquote class="blockquote">
<pre><code> make_var_starts_with (var_starts_with:Optional[Dict[str,tuple]]={})</code></pre>
</blockquote>
<hr>
<p><a href="https://github.com/dsm-72/degex/blob/main/degex/adata.py#L271" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="var_starts_with_pattern" class="level3">
<h3 class="anchored" data-anchor-id="var_starts_with_pattern">var_starts_with_pattern</h3>
<blockquote class="blockquote">
<pre><code> var_starts_with_pattern (name:str, patterns:Optional[Sequence[str]]=None)</code></pre>
</blockquote>
<pre><code>/opt/hostedtoolcache/Python/3.11.4/x64/lib/python3.11/site-packages/fastcore/docscrape.py:225: UserWarning: Unknown section See Also
  else: warn(msg)</code></pre>
<hr>
<p><a href="https://github.com/dsm-72/degex/blob/main/degex/adata.py#L362" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="apply_filter_by_cutoffs" class="level3">
<h3 class="anchored" data-anchor-id="apply_filter_by_cutoffs">apply_filter_by_cutoffs</h3>
<blockquote class="blockquote">
<pre><code> apply_filter_by_cutoffs (adata:anndata._core.anndata.AnnData,
                          cutoff_specs:List[degex.types.CutoffSpec],
                          print_counts:bool=False)</code></pre>
</blockquote>
<p>Uses <code>obs_key</code> to filter <code>adata</code> between <code>lower</code> and <code>upper</code> if provided e.g.&nbsp; <code>lower &lt; adata.obs[obs_key] &lt; upper</code></p>
<table class="table">
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>adata</td>
<td>AnnData</td>
<td></td>
<td>AnnData to process.</td>
</tr>
<tr class="even">
<td>cutoff_specs</td>
<td>List</td>
<td></td>
<td>Specifications in the form of <code>(obs_key, lower, upper)</code><br>to be used to <a href="https://dsm-72.github.io/spot-mark-gene/adata.html#filter_by_cutoffs"><code>filter_by_cutoffs</code></a>.</td>
</tr>
<tr class="odd">
<td>print_counts</td>
<td>bool</td>
<td>False</td>
<td>Whether or not to print counts.</td>
</tr>
<tr class="even">
<td><strong>Returns</strong></td>
<td><strong>AnnData</strong></td>
<td></td>
<td><strong>for function chaining</strong></td>
</tr>
</tbody>
</table>
<hr>
<p><a href="https://github.com/dsm-72/degex/blob/main/degex/adata.py#L312" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="filter_by_cutoffs" class="level3">
<h3 class="anchored" data-anchor-id="filter_by_cutoffs">filter_by_cutoffs</h3>
<blockquote class="blockquote">
<pre><code> filter_by_cutoffs (adata:anndata._core.anndata.AnnData, lower:float=None,
                    upper:float=None, obs_key:Union[ForwardRef('total_coun
                    ts'),ForwardRef('pct_counts_mito'),ForwardRef('pct_cou
                    nts_ribo'),ForwardRef('doublet_scores')]='total_counts
                    ', print_counts:bool=False)</code></pre>
</blockquote>
<p>Uses <code>obs_key</code> to filter <code>adata</code> between <code>lower</code> and <code>upper</code> if provided e.g.&nbsp; <code>lower &lt; adata.obs[obs_key] &lt; upper</code></p>
<table class="table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>adata</td>
<td>AnnData</td>
<td></td>
<td>AnnData to process.</td>
</tr>
<tr class="even">
<td>lower</td>
<td>float</td>
<td>None</td>
<td>Defaults to <code>None</code>. The value which <code>adata.obs[obs_key]</code> <br>should be greater than. If <code>None</code>, <code>lower</code> is not used.</td>
</tr>
<tr class="odd">
<td>upper</td>
<td>float</td>
<td>None</td>
<td>Defaults to <code>None</code>. The value which <code>adata.obs[obs_key]</code> <br>should be less than. If <code>None</code>, <code>upper</code> is not used.</td>
</tr>
<tr class="even">
<td>obs_key</td>
<td>Union</td>
<td>total_counts</td>
<td>Which observation to test against.<br>One of <code>'total_counts'</code>, <code>'pct_counts_mito'</code>, <br><code>'pct_counts_ribo'</code>, or <code>'doublet_score'</code>. Defaults<br>to <code>'total_counts'</code>.</td>
</tr>
<tr class="odd">
<td>print_counts</td>
<td>bool</td>
<td>False</td>
<td>Whether or not to print counts.</td>
</tr>
<tr class="even">
<td><strong>Returns</strong></td>
<td><strong>AnnData</strong></td>
<td></td>
<td><strong>for function chaining</strong></td>
</tr>
</tbody>
</table>
</section>
</section>
<section id="normalization" class="level2">
<h2 class="anchored" data-anchor-id="normalization">Normalization</h2>
<hr>
<p><a href="https://github.com/dsm-72/degex/blob/main/degex/adata.py#L445" target="_blank" style="float:right; font-size:smaller">source</a></p>
<section id="sqrt_library_size_normalize" class="level3">
<h3 class="anchored" data-anchor-id="sqrt_library_size_normalize">sqrt_library_size_normalize</h3>
<blockquote class="blockquote">
<pre><code> sqrt_library_size_normalize (adata:anndata._core.anndata.AnnData)</code></pre>
</blockquote>
<hr>
<p><a href="https://github.com/dsm-72/degex/blob/main/degex/adata.py#L418" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="add_gene_detection_layer" class="level3">
<h3 class="anchored" data-anchor-id="add_gene_detection_layer">add_gene_detection_layer</h3>
<blockquote class="blockquote">
<pre><code> add_gene_detection_layer (adata:anndata._core.anndata.AnnData)</code></pre>
</blockquote>
<hr>
<p><a href="https://github.com/dsm-72/degex/blob/main/degex/adata.py#L400" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="add_prenormalization_layer" class="level3">
<h3 class="anchored" data-anchor-id="add_prenormalization_layer">add_prenormalization_layer</h3>
<blockquote class="blockquote">
<pre><code> add_prenormalization_layer (adata:anndata._core.anndata.AnnData)</code></pre>
</blockquote>
<hr>
<p><a href="https://github.com/dsm-72/degex/blob/main/degex/adata.py#L539" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="load_human_genes" class="level3">
<h3 class="anchored" data-anchor-id="load_human_genes">load_human_genes</h3>
<blockquote class="blockquote">
<pre><code> load_human_genes (adata:anndata._core.anndata.AnnData, filename:str)</code></pre>
</blockquote>
<hr>
<p><a href="https://github.com/dsm-72/degex/blob/main/degex/adata.py#L500" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="score_genes_cell_cycle_with_batch_mean_center_data" class="level3">
<h3 class="anchored" data-anchor-id="score_genes_cell_cycle_with_batch_mean_center_data">score_genes_cell_cycle_with_batch_mean_center_data</h3>
<blockquote class="blockquote">
<pre><code> score_genes_cell_cycle_with_batch_mean_center_data
                                                     (adata:anndata._core.
                                                     anndata.AnnData, s_ge
                                                     nes:Sequence[str], g2
                                                     m_genes:Sequence[str]
                                                     )</code></pre>
</blockquote>
<hr>
<p><a href="https://github.com/dsm-72/degex/blob/main/degex/adata.py#L472" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="add_batch_mean_center_layer" class="level3">
<h3 class="anchored" data-anchor-id="add_batch_mean_center_layer">add_batch_mean_center_layer</h3>
<blockquote class="blockquote">
<pre><code> add_batch_mean_center_layer (adata:anndata._core.anndata.AnnData)</code></pre>
</blockquote>
</section>
</section>
<section id="highly-variable-genes" class="level2">
<h2 class="anchored" data-anchor-id="highly-variable-genes">Highly Variable Genes</h2>
<hr>
<p><a href="https://github.com/dsm-72/degex/blob/main/degex/adata.py#L574" target="_blank" style="float:right; font-size:smaller">source</a></p>
<section id="select_hvg_per_batch" class="level3">
<h3 class="anchored" data-anchor-id="select_hvg_per_batch">select_hvg_per_batch</h3>
<blockquote class="blockquote">
<pre><code> select_hvg_per_batch (adata:anndata._core.anndata.AnnData,
                       hvg_kwargs:dict={'cutoff': None, 'percentile': 90})</code></pre>
</blockquote>
<p>Calculates highly variable genes per batch in <code>adata</code></p>
<table class="table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>adata</td>
<td>AnnData</td>
<td></td>
<td>AnnData to process.</td>
</tr>
<tr class="even">
<td>hvg_kwargs</td>
<td>dict</td>
<td>{‘cutoff’: None, ‘percentile’: 90}</td>
<td>Options to be passed to <code>sc.select_highly_variable_genes</code>.</td>
</tr>
<tr class="odd">
<td><strong>Returns</strong></td>
<td><strong>AnnData</strong></td>
<td></td>
<td><strong>for function chaining</strong></td>
</tr>
</tbody>
</table>
</section>
</section>
<section id="annotations-1" class="level2">
<h2 class="anchored" data-anchor-id="annotations-1">Annotations</h2>
<hr>
<p><a href="https://github.com/dsm-72/degex/blob/main/degex/adata.py#L637" target="_blank" style="float:right; font-size:smaller">source</a></p>
<section id="add_mouse_tfs_from_csv" class="level3">
<h3 class="anchored" data-anchor-id="add_mouse_tfs_from_csv">add_mouse_tfs_from_csv</h3>
<blockquote class="blockquote">
<pre><code> add_mouse_tfs_from_csv (adata:anndata._core.anndata.AnnData,
                         filename:str, print_counts:bool=False)</code></pre>
</blockquote>
<hr>
<p><a href="https://github.com/dsm-72/degex/blob/main/degex/adata.py#L628" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="add_human_tfs_from_csv" class="level3">
<h3 class="anchored" data-anchor-id="add_human_tfs_from_csv">add_human_tfs_from_csv</h3>
<blockquote class="blockquote">
<pre><code> add_human_tfs_from_csv (adata:anndata._core.anndata.AnnData,
                         filename:str, print_counts:bool=False)</code></pre>
</blockquote>
<hr>
<p><a href="https://github.com/dsm-72/degex/blob/main/degex/adata.py#L613" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="add_tf_annotations_from_csv" class="level3">
<h3 class="anchored" data-anchor-id="add_tf_annotations_from_csv">add_tf_annotations_from_csv</h3>
<blockquote class="blockquote">
<pre><code> add_tf_annotations_from_csv (adata:anndata._core.anndata.AnnData,
                              filename:str, tf_key:str, ensemble_key:str,
                              print_counts:bool=False)</code></pre>
</blockquote>
</section>
</section>
<section id="normalization-1" class="level2">
<h2 class="anchored" data-anchor-id="normalization-1">Normalization</h2>
<hr>
<p><a href="https://github.com/dsm-72/degex/blob/main/degex/adata.py#L666" target="_blank" style="float:right; font-size:smaller">source</a></p>
<section id="subset_markers" class="level3">
<h3 class="anchored" data-anchor-id="subset_markers">subset_markers</h3>
<blockquote class="blockquote">
<pre><code> subset_markers (adata:anndata._core.anndata.AnnData,
                 obs_key:str='Markers_cell',
                 score_key:str='Markers_zscore', lower:float=2.2,
                 upper:float=None, marker_name:str='marker',
                 other_name:str='other')</code></pre>
</blockquote>
<hr>
<p><a href="https://github.com/dsm-72/degex/blob/main/degex/adata.py#L649" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="zscore_markers_in_layer" class="level3">
<h3 class="anchored" data-anchor-id="zscore_markers_in_layer">zscore_markers_in_layer</h3>
<blockquote class="blockquote">
<pre><code> zscore_markers_in_layer (adata:anndata._core.anndata.AnnData,
                          markers:List[str], obs_key:str='Markers_zscore',
                          layer_key:str='X_magic')</code></pre>
</blockquote>
</section>
</section>
<section id="embeddings" class="level2">
<h2 class="anchored" data-anchor-id="embeddings">Embeddings</h2>
<section id="pca" class="level3">
<h3 class="anchored" data-anchor-id="pca">PCA</h3>
<hr>
<p><a href="https://github.com/dsm-72/degex/blob/main/degex/adata.py#L718" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="run_pca_on_hvg" class="level3">
<h3 class="anchored" data-anchor-id="run_pca_on_hvg">run_pca_on_hvg</h3>
<blockquote class="blockquote">
<pre><code> run_pca_on_hvg (adata:anndata._core.anndata.AnnData,
                 pca_kwargs:dict={'n_components': 100},
                 plot_scree:bool=False)</code></pre>
</blockquote>
<hr>
<p><a href="https://github.com/dsm-72/degex/blob/main/degex/adata.py#L696" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="run_pca" class="level3">
<h3 class="anchored" data-anchor-id="run_pca">run_pca</h3>
<blockquote class="blockquote">
<pre><code> run_pca (adata:anndata._core.anndata.AnnData,
          pca_kwargs:dict={'n_components': 100}, plot_scree:bool=False,
          emb_key:str='X_pca', col_subset:Union[list,pandas.core.series.Se
          ries,numpy.ndarray,Iterable[numbers.Number],Any,List[int],List[U
          nion[bool,numpy.bool_,Literal[0],Literal[1]]],pandas.core.indexe
          s.base.Index]=None)</code></pre>
</blockquote>
</section>
<section id="phate" class="level3">
<h3 class="anchored" data-anchor-id="phate">PHATE</h3>
<hr>
<p><a href="https://github.com/dsm-72/degex/blob/main/degex/adata.py#L756" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="run_phate_on_hvg" class="level3">
<h3 class="anchored" data-anchor-id="run_phate_on_hvg">run_phate_on_hvg</h3>
<blockquote class="blockquote">
<pre><code> run_phate_on_hvg (adata:anndata._core.anndata.AnnData, g:&lt;function
                   Graph&gt;=None, phate_kwargs:dict={'n_components': 3, 't':
                   70}, g_kwargs:dict={'knn': 10},
                   emb_key:str='X_phate_hvg')</code></pre>
</blockquote>
<hr>
<p><a href="https://github.com/dsm-72/degex/blob/main/degex/adata.py#L727" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="run_phate_using_g" class="level3">
<h3 class="anchored" data-anchor-id="run_phate_using_g">run_phate_using_g</h3>
<blockquote class="blockquote">
<pre><code> run_phate_using_g (adata:anndata._core.anndata.AnnData, g:&lt;function
                    Graph&gt;=None, phate_kwargs:dict={'n_components': 3,
                    't': 70}, g_kwargs:dict={'knn': 10},
                    emb_key:str='X_phate')</code></pre>
</blockquote>
</section>
<section id="magic" class="level3">
<h3 class="anchored" data-anchor-id="magic">MAGIC</h3>
<hr>
<p><a href="https://github.com/dsm-72/degex/blob/main/degex/adata.py#L769" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="run_magic" class="level3">
<h3 class="anchored" data-anchor-id="run_magic">run_magic</h3>
<blockquote class="blockquote">
<pre><code> run_magic (adata:anndata._core.anndata.AnnData, g:&lt;function Graph&gt;,
            magic_kwargs:dict={'solver': 'approximate', 'n_jobs': -1,
            'knn_max': 60})</code></pre>
</blockquote>


</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>